TEST: Add tests against local minio.

ssanderson opened the pull request.

Commit:
TEST: Add tests against local minio.

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+    pid_t m_pid;
+
+public:
+    process(std::vector<std::string>& args,
```
should these be const?

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+        for (auto& str : m_envp) {
+            c_env.emplace_back(str.data());
+        }
+        c_env.push_back(NULL);
```
`nullptr`

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+        c_args.push_back(NULL);
+
+        // Initialize envp.
+        for (unsigned int i = 0; i < env.size(); ++i) {
```
std::size_t is normally used for iteration variables

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+            environment& env) {
+
+        // Initialize argv.
+        for (unsigned int i = 0; i < args.size(); ++i) {
```
should we just use the vector copy constructor in the initializer list for this ctor?

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+        explicit error(const std::string& message) : std::runtime_error(message) {}
+    };
+
+private:
```
Can we keep the visibility groups contiguous where possible?

ssanderson replied:
These were originally splt b/c I was using the `environment` typedef in the private section, but I wanted to make it public.  I'm not using that anymore, so I consolidate these.

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+
+namespace h5s3::testing {
+
+class process {
```
can we get some javadoc comments on this? kv_driver has some examples

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+
+
+        // fork/exec
+        pid_t result = fork();
```
should we just set m_pid directly?

llllllllll commented on tests/test_process.cc:
```diff
@@ -0,0 +1,77 @@
+}
+
+TEST(process, interrupt) {
+    std::vector<std::string> args = {"python", "-c", "while True:\n    pass"};
```
If this doesn't actually work, won't we be stuck forever? What if we just ran something like `sleep` so that we can at least continue execution.

llllllllll commented on tests/test_process.cc:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    // Joining for a second time should trigger an error.
+    EXPECT_THROW({ proc.join(); }, process::error);
```
TIL EXPECT_THROW

llllllllll commented on tests/process.h:
```diff
@@ -0,0 +1,118 @@
+        }
+    }
+
+    void terminate() {
```
should the dtor call this?

llllllllll commented on tests/test_s3.cc:
```diff
@@ -0,0 +1,94 @@
+            {"MINIO_SECRET_KEY", SECRET_KEY},
+            {"PATH", std::getenv("PATH")},
+        };
+        minio_proc = std::make_unique<process>(args, env);
```
why is this heap allocated instead of just a value?

ssanderson replied:
talked about this in person.  we're using the unique pointer to allow the lifetime of the process to be shorter than the lifetime of the test suite.

llllllllll commented on tests/test_s3.cc:
```diff
@@ -0,0 +1,94 @@
+            "remove",
+            "h5s3-testing",
+        };
+        process::environment empty_env = {{"PATH", std::getenv("PATH")}};
```
do we actually need the path? what is it looking for?

ssanderson replied:
minio shells out to something on startup.  without this I was getting errors from minio failing to find /bin/sh.

llllllllll commented on tests/test_s3.cc:
```diff
@@ -0,0 +1,94 @@
+
+
+TEST_F(S3Test, test_set_get) {
+    s3::notary signer("us-east-1", ACCESS_KEY, SECRET_KEY);
```
should the test fixture provide a notary?

llllllllll commented on tests/test_s3.cc:
```diff
@@ -0,0 +1,94 @@
+TEST_F(S3Test, test_set_get) {
+    s3::notary signer("us-east-1", ACCESS_KEY, SECRET_KEY);
+    s3::set_object(signer, "some-bucket", "foo/bar", "my-content", "localhost:9999", false);
+    std::string result = s3::get_object(signer, "some-bucket", "foo/bar", "localhost:9999", false);
```
should the fixture provide the host?

llllllllll commented on etc/install-minio.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+download_url="https://dl.minio.io/server/minio/release/linux-amd64/minio"
```
let's just put this in the Makefile as a dependency of the tests. I did that for gtest in libpy and it seemed pretty nice.

llllllllll reviewed (commented):
Overall this looks pretty sweet!

Commit:
TEST: Clean up s3 tests.

- Dedupe common command code into a lambda.
- Test multiple set/get pairs.
- Add test for 404.

Commit:
MAINT: Add Makefile target for clang-tidy.

Commit:
MAINT: Add missing #pragma: once lines.

Commit:
TEST: More test updates.

- Add docs to `process`.
- Move minio install script to Makefile.
- Call `std::terminate` if process isn't joined.

Commit:
BLD: Add Makefile options for running with ubsan.

Commit:
BLD/TEST: Test files depend on .compiler_flags.

Commit:
DOC: Fix typo.
